### ISSUE TITLE
`fs.gs.glob.flatlist.enable = true` is an unsafe default value

### ISSUE THREAD

**Issue by srosenberg** [GoogleCloudPlatform/bigdata-interop#106]
If a bucket contains a large number of blobs, flat-glob can cause OOM which is exactly what happened to our production spark jobs.  Default values should be chosen judiciously.  This is a case where the default is aggressive (i.e., potentially improves performance) as opposed to conservative.

**Comment by medb** [comment]
May you provide more details? What GCS connector version did you use? How many objects do you have in the bucket that caused OOM during flat globing?

**Comment by srosenberg** [comment]
@medb The exception resulted from Hadoop GCS connector which is used in conjunction with dataproc.  The version is `1.6.7` based on the loaded jar: `gcs-connector-1.6.7-hadoop2.jar`.  The partial stack trace is below,

```
java.lang.OutOfMemoryError: GC overhead limit exceeded
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at java.util.Arrays.copyOfRange(Arrays.java:3664)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at java.lang.String.<init>(String.java:207)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at java.lang.String.substring(String.java:1969)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at java.net.URI$Parser.substring(URI.java:2869)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at java.net.URI$Parser.parseHierarchical(URI.java:3106)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at java.net.URI$Parser.parse(URI.java:3053)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at java.net.URI.<init>(URI.java:746)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at org.apache.hadoop.fs.Path.initialize(Path.java:251)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at org.apache.hadoop.fs.Path.<init>(Path.java:245)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at org.apache.hadoop.fs.Path.getParent(Path.java:431)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at com.google.cloud.hadoop.fs.gcs.InMemoryGlobberFileSystem.<init>(InMemoryGlobberFileSystem.java:86)
29-06-2018 22:00:40 EDT prebid_enrichment_job INFO -     at com.google.cloud.hadoop.fs.gcs.InMemoryGlobberFileSystem.createInstance(InMemoryGlobberFileSystem.java:61
```

At this time, the input bucket contains `> 1 mil.` blobs; the prefix path that is globbed for this job contains `838,311` blobs.

**Comment by medb** [comment]
Thanks for information!

Unfortunately, we are not able to change default behavior, because it could break or negatively impact existing customers.

But, I think that we can improve flat globing to avoid OOMs in such cases. Currently we fetch list of all objects in memory and filter them afterwards, but if we will do this in batches (GCS anyways lists objects in 1,000 pages) we will be able to avoid OOMs without sacrificing performance.

May you provide blobs paths structure, specifically the "depth" of blobs path (i.e. how many subfolders starting from prefix in blobs path that you are globing)?

Also, maybe you had a change to disable flat globing, if so, how big performance hit did you observe, if any?

**Comment by srosenberg** [comment]
@medb The path structure of the blobs is of this form `gs://logs/log_type/source/2018/07/03/23`.  After we disabled flat globing, we noticed a boost in performance; prior to the change, spark jobs would spend ~3 minutes on preparing input whereas now it's a matter of _seconds_.

Using an iterator to page through the results (i.e., streaming) instead of materializing the full result set in memory is indeed a better approach; I've implemented it in a tool that manages retention policies in gcs on a per-path basis rather than per-bucket.  As an example, the following snippet implements a (recursive) blob enumerator and exposes the result as `Iterable<FileStatus>` which can be easily composed with other hadoop/spark path filters,

```
public static class GcsFileStatusEnumerator implements FileStatusEnumerator {
        private Storage storage;

        private static final Storage.BlobListOption BLOB_FIELDS = Storage.BlobListOption.fields(
                Storage.BlobField.STORAGE_CLASS, Storage.BlobField.UPDATED,
                Storage.BlobField.TIME_DELETED, Storage.BlobField.SIZE);

        public GcsFileStatusEnumerator(Storage storage) {
            this.storage = storage;
        }

        @Override
        public Iterable<? extends FileStatus> listStatus(Path path, Filters.FileStatusFilter fileStatusFilter, boolean recursive) {
            URI uri = path.toUri();
            Preconditions.checkArgument(uri.getScheme().equalsIgnoreCase("gs"), "non-gcs path: %s", path);

            String bucketName = uri.getAuthority();
            String objectName = uri.getPath();
            // removing leading '/', if it exists
            if (objectName.startsWith("/")) {
                objectName = objectName.substring(1);
            }
            List<FileStatus> result = new ArrayList<>();
            Storage.BlobListOption[] blobListOptions;

            if (!recursive) {
                blobListOptions = new Storage.BlobListOption[] {
                        Storage.BlobListOption.prefix(objectName), Storage.BlobListOption.currentDirectory(), BLOB_FIELDS
                };
            } else {
                blobListOptions = new Storage.BlobListOption[] {
                        Storage.BlobListOption.prefix(objectName), BLOB_FIELDS
                };
            }
            Iterable<Blob> allBlobs = storage.list(bucketName, blobListOptions).iterateAll();

            return FluentIterable.from(allBlobs).transform(blob -> new GcsFilesStatus(blob)).filter(fileStatus -> fileStatusFilter.accept(fileStatus));
        }
    }
```

**Comment by medb** [comment]
Very interesting, it could be the case that after [recent change](https://github.com/GoogleCloudPlatform/bigdata-interop/commit/e5c6a5f429cce2b60733f2ec43d4488b4fb19e91) where GCS connector took advantage of GCS `includeTrailingDelimiter` feature non-flat globs become faster than flat globs.

Need to validate this, and if this is the case, we could change default.

Another idea for posterity:
run 1 flat-glob request (max 1,000 objects) and 1 non flat glob requests in parallel, and after receiving responses make decision with what glob logic to proceed - if prefix has 1,000 objects or less then use flat-glob (all objects info already fetched in 1st request), if no - continue with non-flat globing.

**Comment by yzhou2001** [comment]
I run glob queries on a big GCS object "tree" of about 890K entries. Following are the observations and comments:

1) There is no significant performance discrepancy between using includeTrailingDelimiter or not, with glob flat disabled. For a small selected result set, 100,  out of the big tree, 890K, the wall clock time is 10 seconds vs 11 seconds;
2) when glob flat is enabled, the performance is much slowed down indeed. The issue is that a most permissive filtering is passed to GCS, resulting in huge amount of candidate objects returned to the connector. For instance, a query of gs://bucketabc/1???/1/1/small-1.txt causes gs://yazhou_reddit1/1 to be used as a prefix passed to GCS. This has caused the significant performance slow down. I have filed another issue for this: https://github.com/GoogleCloudPlatform/bigdata-interop/issues/123.
3) when glob flat is enabled, the use of the most permissive filtering on the GCS stores a big candidate set before a "glob pattern" is used to select the final results against the big candidate set.
4) Depending upon the lengths of the fully qualified object paths, the HDFS client app, and its jvm heap limit setting, it may or may not burst with the OOM. I'm using objects with about 40-char(80-byte) full lengths, and Dataproc default JVM heap settings. Using Hadoop, the resident mem raised up to 3.3GB;
using Spark, the resident mem raised up to 4.0GB. The reason being that Spark in its relation resolution makes 2 "glob status" calls in a quick succession that almost double the memory pool to hold the big "candidate" entries. However, I did not actually hit the OOM.
5) During the shipping of the "candidate" entries from GFS, the network traffic is only about 40Mb/s, way below the capacity. Running a second client in parallel immediately boosts the network transfer rate to about 1.5X, confirming that the resources are underused. As result, the connector spends mostly waiting for the incoming data. This could have been improved in 2 ways: i) bump up the batch size limit. Currently it's hard set at 1000; in contrast AWS has 5000 as default( fs.s3a.paging.maximum); and ii) parallelization of shipping the qualified entries. (AWS connector does not have this parallelization either). The first would require changes at GCS-side. The second had better to be supported by GCS. But with big data processing more popular agains GCS, I expect that it will become ever more important.

A suggested fix of the excessive memory usage is to use a "glob filter" to filter the "candidate" set "in-place" per page before save it to the final result. A prototype has shown the resident memory usage down to 300M from 3.3G using Hadoop, and 1.7G from 4.0G using Spark, with about the same speed.

I believe the "glob flat" optimization makes more sense with deeper and/or wider "trees". So a fix should be desirable.

**Comment by medb** [comment]
Thanks for investigation!

Unfortunately GCS does not return more than 1,000 objects in one response: https://cloud.google.com/storage/docs/json_api/v1/objects/list

Yes, glob flat optimization designed to address different issue and it works for the case of deep directory tries.

After suggested fix with globbing each result page, we can consider optimization with running flat and non-flat globs in parallel and using results of whichever is complete first.

**Comment by dennishuo** [comment]
1000 should be reasonably performant. Even though the S3 client requests 5000, it can only get 1000 per page still: https://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html https://github.com/aws/aws-sdk-cpp/issues/369

**Comment by yzhou2001** [comment]
S3's 1000 seems to be the default and not the hard ceiling as that of  GCS. I tried to scale down 1000 to 500 and 250. It appears that the performance kinda stabilizes after 500, so 1000 is probably good enough now. But the fundamental limiting factor is at the server side while the client keeps waiting for data from GCS.  And client side settings will have to be dependent upon server behavior and will deserve a revisit once server performance can be significantly improved.

I'd like to call out again for the importance of this improvement consideration, given the scalability and performance requirements of modern big data processing technologies. That is, many thousands of processing tasks plus numerous multi-level data partitions have made "online" querying over a big data tree ever more common. There have been many complains against popular big data engines when they can't scale well as claimed. It definitely makes sense to proactively fix issues as perceived.

**Comment by medb** [comment]
@srosenberg May you share a glob pattern that you mentioned above - executes in seconds with regular globbing and ~3 minutes with flat globbing? It will be useful for testing further optimizations to our globbing algorithm.

**Comment by srosenberg** [comment]
@medb  Here is an example: `gs://logs/log_type/*/2018/10/14/07`.  The globbed expression matches all sources of a given log type; in our case, there are 3 possible sources.  (However, as I mentioned earlier, `logs` bucket contains a very large number of blobs.)

**Comment by medb** [comment]
@srosenberg Thank you, this is very helpful! I have reproduced the issue on a test dataset and got the same results as you (seconds vs minutes) - we will work to address this issue in default configuration.

**Comment by medb** [comment]
@srosenberg it was closed because we [fixed](https://github.com/GoogleCloudPlatform/bigdata-interop/commit/341a53920144921aa973c50fe5447af8ee49205c) OOMs in flat glob in latest connector version (1.9.8). But we didn't improve runtime, so I still would like to follow up with performance optimizations that are tracked as a separate [issue](#123).
